File: src/test/java/com/github/jactor/persistence/aop/ModifierAspectTest.java
package com.github.jactor.persistence.aop;

import static com.github.jactor.persistence.entity.AddressEntity.anAddress;
import static com.github.jactor.persistence.entity.BlogEntity.aBlog;
import static com.github.jactor.persistence.entity.BlogEntryEntity.aBlogEntry;
import static com.github.jactor.persistence.entity.GuestBookEntity.aGuestBook;
import static com.github.jactor.persistence.entity.GuestBookEntryEntity.aGuestBookEntry;
import static com.github.jactor.persistence.entity.PersonEntity.aPerson;
import static com.github.jactor.persistence.entity.UserEntity.aUser;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import com.github.jactor.persistence.dto.AddressDto;
import com.github.jactor.persistence.dto.BlogDto;
import com.github.jactor.persistence.dto.BlogEntryDto;
import com.github.jactor.persistence.dto.GuestBookDto;
import com.github.jactor.persistence.dto.GuestBookEntryDto;
import com.github.jactor.persistence.dto.PersistentDto;
import com.github.jactor.persistence.dto.PersonInternalDto;
import com.github.jactor.persistence.dto.UserInternalDto;
import java.time.LocalDateTime;
import org.aspectj.lang.JoinPoint;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
@DisplayName("ModifierAspect")
class ModifierAspectTest {

  private LocalDateTime oneMinuteAgo = LocalDateTime.now().minusMinutes(1);
  private ModifierAspect modifierAspect = new ModifierAspect();
  private PersistentDto persistentDto = new PersistentDto(null, "na", oneMinuteAgo, "na", oneMinuteAgo);

  @Mock
  private JoinPoint joinPointMock;


  @Test
  @DisplayName("should modify timestamp on address when used")
  void shouldModifyTimestampOnAddressWhenUsed() {
    var addressWithoutId = anAddress(new AddressDto(persistentDto, new AddressDto()));
    var address = anAddress(new AddressDto(persistentDto, new AddressDto()));
    address.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{address, addressWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(address.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(addressWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on blog when used")
  void shouldModifyTimestampOnBlogWhenUsed() {
    var blogWithouId = aBlog(new BlogDto(persistentDto, new BlogDto()));
    var blog = aBlog(new BlogDto(persistentDto, new BlogDto()));
    blog.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{blog, blogWithouId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(blog.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(blogWithouId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on blogEntry when used")
  void shouldModifyTimestampOnBlogEntryWhenUsed() {
    var blogEntryWithoutId = aBlogEntry(new BlogEntryDto(persistentDto, new BlogEntryDto()));
    var blogEntry = aBlogEntry(new BlogEntryDto(persistentDto, new BlogEntryDto()));
    blogEntry.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{blogEntry, blogEntryWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(blogEntry.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(blogEntryWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on guestBook when used")
  void shouldModifyTimestampOnGuestBookWhenUsed() {
    var guestBookWithoutId = aGuestBook(new GuestBookDto(persistentDto, new GuestBookDto()));
    var guestBook = aGuestBook(new GuestBookDto(persistentDto, new GuestBookDto()));
    guestBook.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{guestBook, guestBookWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(guestBook.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(guestBookWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on guestBookEntry when used")
  void shouldModifyTimestampOnGuestBookEntryWhenUsed() {
    var guestBookEntryWithoutId = aGuestBookEntry(new GuestBookEntryDto(persistentDto, new GuestBookEntryDto()));
    var guestBookEntry = aGuestBookEntry(new GuestBookEntryDto(persistentDto, new GuestBookEntryDto()));
    guestBookEntry.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{guestBookEntry, guestBookEntryWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(guestBookEntry.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(guestBookEntryWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on person when used")
  void shouldModifyTimestampOnPersonWhenUsed() {
    var personWithoutId = aPerson(new PersonInternalDto(persistentDto, new PersonInternalDto()));
    var person = aPerson(new PersonInternalDto(persistentDto, new PersonInternalDto()));
    person.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{person, personWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(person.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(personWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }

  @Test
  @DisplayName("should modify timestamp on user when used")
  void shouldModifyTimestampOnUserWhenUsed() {
    var userWithoutId = aUser(new UserInternalDto(persistentDto, new UserInternalDto()));
    var user = aUser(new UserInternalDto(persistentDto, new UserInternalDto()));
    user.setId(1L);

    when(joinPointMock.getArgs()).thenReturn(new Object[]{user, userWithoutId});
    modifierAspect.modifyPersistentEntity(joinPointMock);

    assertThat(user.getTimeOfModification()).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now());
    assertThat(userWithoutId.getTimeOfModification()).isEqualTo(oneMinuteAgo);
  }
}

File: src/test/kotlin/com/github/jactor/persistence/aop/ModifierAspectTest.kt
package com.github.jactor.persistence.aop

import com.github.jactor.persistence.dto.AddressInternalDto
import com.github.jactor.persistence.dto.BlogDto
import com.github.jactor.persistence.dto.BlogEntryDto
import com.github.jactor.persistence.dto.GuestBookDto
import com.github.jactor.persistence.dto.GuestBookEntryDto
import com.github.jactor.persistence.dto.PersistentDto
import com.github.jactor.persistence.dto.PersonInternalDto
import com.github.jactor.persistence.dto.UserInternalDto
import com.github.jactor.persistence.entity.AddressEntity.Companion.anAddress
import com.github.jactor.persistence.entity.BlogEntity.Companion.aBlog
import com.github.jactor.persistence.entity.BlogEntryEntity.Companion.aBlogEntry
import com.github.jactor.persistence.entity.GuestBookEntity.Companion.aGuestBook
import com.github.jactor.persistence.entity.GuestBookEntryEntity.Companion.aGuestBookEntry
import com.github.jactor.persistence.entity.PersonEntity.Companion.aPerson
import com.github.jactor.persistence.entity.UserEntity.Companion.aUser
import org.aspectj.lang.JoinPoint
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
internal class ModifierAspectTest {
    private val oneMinuteAgo = LocalDateTime.now().minusMinutes(1)
    private val modifierAspect = ModifierAspect()
    private val persistentDto = PersistentDto(null, "na", oneMinuteAgo, "na", oneMinuteAgo)

    @Mock
    private val joinPointMock: JoinPoint? = null

    @Test
    fun `should modify timestamp on address when used`() {
        val addressWithoutId = anAddress(AddressInternalDto(persistentDto, AddressInternalDto()))
        val address = anAddress(AddressInternalDto(persistentDto, AddressInternalDto()))
        address.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(address, addressWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(address.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(addressWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on blog when used`() {
        val blogWithouId = aBlog(BlogDto(persistentDto, BlogDto()))
        val blog = aBlog(BlogDto(persistentDto, BlogDto()))
        blog.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(blog, blogWithouId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(blog.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(blogWithouId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on blogEntry when used`() {
        val blogEntryWithoutId = aBlogEntry(BlogEntryDto(persistentDto, BlogEntryDto()))
        val blogEntry = aBlogEntry(BlogEntryDto(persistentDto, BlogEntryDto()))
        blogEntry.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(blogEntry, blogEntryWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(blogEntry.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(blogEntryWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on guestBook when used`() {
        val guestBookWithoutId = aGuestBook(GuestBookDto(persistentDto, GuestBookDto()))
        val guestBook = aGuestBook(GuestBookDto(persistentDto, GuestBookDto()))
        guestBook.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(guestBook, guestBookWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(guestBook.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(guestBookWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on guestBookEntry when used`() {
        val guestBookEntryWithoutId = aGuestBookEntry(GuestBookEntryDto(persistentDto, GuestBookEntryDto()))
        val guestBookEntry = aGuestBookEntry(GuestBookEntryDto(persistentDto, GuestBookEntryDto()))
        guestBookEntry.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(guestBookEntry, guestBookEntryWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(guestBookEntry.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(guestBookEntryWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on person when used`() {
        val personWithoutId = aPerson(PersonInternalDto(persistentDto, PersonInternalDto()))
        val person = aPerson(PersonInternalDto(persistentDto, PersonInternalDto()))
        person.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(person, personWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(person.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(personWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }

    @Test
    fun `should modify timestamp on user when used`() {
        val userWithoutId = aUser(UserInternalDto(persistentDto, UserInternalDto()))
        val user = aUser(UserInternalDto(persistentDto, UserInternalDto()))
        user.id = 1L
        Mockito.`when`(joinPointMock!!.args).thenReturn(arrayOf<Any>(user, userWithoutId))
        modifierAspect.modifyPersistentEntity(joinPointMock)
        Assertions.assertThat(user.timeOfModification).isStrictlyBetween(LocalDateTime.now().minusSeconds(1), LocalDateTime.now())
        Assertions.assertThat(userWithoutId.timeOfModification).isEqualTo(oneMinuteAgo)
    }
}